src/std/net/s3/api.ss

```diff
@@ -0,0 +1,231 @@
+                  (code (request-status req)))
+             ; 200 and 404 are expected codes
+             ; we explicitly handle 404 so we get proper predicate
+             ; semantics and don't raise on what would otherwise be
+             ; #f condition.
+             (if (memv code [200 404]) 
+               (begin
+                 (request-close req)
+                 (= code 200))
+               (with-request-error req))))))
+
+(defmethod {bucket s3-client}
+  (lambda (self name)
+    (using (self self :- s3-client)
+      (if (s3-client::bucket-exists? self name)
+        (make-bucket self name (s3-client-region self))
+        #f))))
+
+; Lists the objects stored within the bucket
+(defmethod {list-objects bucket}
+  (lambda (self)
+    (using ((self self :- bucket)
+            (client (bucket-client self) : s3-client))
+           (let* ((name (bucket-name self))
+                  (req (s3-request/error client verb: 'GET bucket: name))
+                  (xml (s3-parse-xml req))
+                  (keys (sxml-select xml (sxml-e? 's3:Key) cadr)))
+             (request-close req)
+             keys))))
+
+(defmethod {get bucket}
+  (lambda (self key)
+    (using ((self self : bucket)
+            (client (bucket-client self) : s3-client))
+      (let* ((req (s3-request/error client verb: 'GET bucket: (bucket-name self)
+                                    path: (string-append "/" key)))
+             (data (request-content req)))
+        (request-close req)
+        data))))
+
+(defmethod {put! bucket}
+  (lambda (self key data content-type: (content-type "binary/octet-stream"))
+    (using ((self self : bucket)
+            (client (bucket-client self) : s3-client))
+           (let (req (s3-request/error client verb: 'PUT bucket: (bucket-name self)
+                                       path: (string-append "/" key)
+                                       body: data
+                                       content-type: content-type))
+             (request-close req)
+             (void)))))
+
+(defmethod {delete! bucket}
+  (lambda (self key)
+    (using ((self :- bucket)
+            (client (bucket-client self) : s3-client))
+           (let (req (s3-request/error client verb: 'DELETE bucket: (bucket-name self)
+                     path: (string-append "/" key)))
+             (request-close req)
+             (void)))))
+
+(defmethod {copy-to! bucket}
+  (lambda (self src dest)
+    (using ((self :- bucket)
+            (client (bucket-client self) : s3-client))
+           (let* ((headers [["x-amz-copy-source" :: src]])
+                  (req (s3-request/error client
+                                         verb: 'PUT
+                                         bucket: (bucket-name self)
+                                         path: (string-append "/" dest)
+                                         extra-headers: headers)))
+             (request-close req)
+             (void)))))
```

Comment:
The docs say that a copy can return a 200 response but report a failure in the message body. I'll have to do some extra work to identify what the failure message looks like and parse it out

Reply:
I think I handled this ok now